### PR TITLE
Configurable training

### PR DESCRIPTION
**Pull request by saranggalada**
1. Added two new scenarios: BraTS and CIFAR-10
2. Pytrain now allows for highly flexible and secure training pipeline configurations
3. Fixed bugs in training code and ACI deployment
4. Restructured scenario subdirectories to demarcate different parties and components

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,17 +44,22 @@ This scripts build the following containers.
 
 Alternatively, you can use pre-built container images from the ispirt repository by setting the following environment variable. Docker hub has started throttling which may effect the upload/download time, especially when images are bigger size. So, It is advisable to use other container registries, we are using azure container registry as shown below
 ```bash
-export CONTAINER_REGISTRY=ispirt.azurecr.io
+export CONTAINER_REGISTRY=depatraindevacr.azurecr.io
```
Comment (kapilvgit) [inline README.md]: suggest we use the ispirt repo

File: ci/Dockerfile.train [hunk 2]
```diff
@@ -1,23 +1,33 @@
...
-RUN python3.9 get-pip.py
+RUN python3 get-pip.py
 
 ## Install dependencies
 RUN pip3 install torch torchvision --index-url https://download.pytorch.org/whl/cpu
-RUN pip3 --default-timeout=1000 install pyspark pandas opacus onnx onnx2pytorch scikit-learn scipy matplotlib 
+RUN pip3 --default-timeout=1000 install pyspark pandas opacus==1.5.3 onnx onnx2pytorch scikit-learn scipy matplotlib 
+RUN pip3 install safetensors h5py pyarrow
+
+## ADDED: For language tasks
```
Comment (kapilvgit) [inline ci/Dockerfile.train]: Remove ADDED

File: scenarios/brats/config/consolidate_pipeline.sh
Comment (kapilvgit) [inline scenarios/brats/config/consolidate_pipeline.sh:1]: At some point we should consider a framework like Hydra for all configuration

File: scenarios/brats/config/model_config_old.json
Comment (kapilvgit) [inline scenarios/brats/config/model_config_old.json:1]: Do we still need these files?

File: scenarios/brats/deployment/azure/deploy.sh [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  # Obtain the token based on the AKV resource endpoint subdomain
+  if [[ "$AZURE_KEYVAULT_ENDPOINT" == *".vault.azure.net" ]]; then
+      export BEARER_TOKEN=$(az account get-access-token --resource https://vault.azure.net | jq -r .accessToken)
+      echo "Importing keys to AKV key vaults can be only of type RSA-HSM"
+      export AZURE_AKV_KEY_TYPE="RSA-HSM"
+  elif [[ "$AZURE_KEYVAULT_ENDPOINT" == *".managedhsm.azure.net" ]]; then
+      export BEARER_TOKEN=$(az account get-access-token --resource https://managedhsm.azure.net | jq -r .accessToken)
+      export AZURE_AKV_KEY_TYPE="oct-HSM"
+  fi
+
```
Comment (kapilvgit) [inline scenarios/brats/deployment/azure/deploy.sh:75]: We should try and clean this up using a loop to go over all datasets and key combinations

File: scenarios/cifar10/config/model_config.json
Comment (kapilvgit) [inline scenarios/cifar10/config/model_config.json:1]: for this scenario, can we not use ONNX?

File: scenarios/brats/deployment/azure/deploy.sh (hunk 3 above)
Comment (saranggalada) [reply scenarios/brats/deployment/azure/deploy.sh:75]: Yes, in a future PR we would ideally create a "Create your own scenario" guide, as part of which the repeated code for deployment is auto-generated from template. This would remove all the hardcoding we have done within these scripts, and factor in number of datasets, etc

File: scenarios/cifar10/config/model_config.json
Comment (saranggalada) [reply scenarios/cifar10/config/model_config.json:1]: Yes, we could. Since it is very similar to the MNIST scenario, I thought we could demonstrate some variability/flexibility by changing model and dataset formats and showing that they work smoothly.

File: scenarios/brats/config/model_config_old.json
Comment (saranggalada) [reply scenarios/brats/config/model_config_old.json:1]: Oops, forgot to delete

**Comment by saranggalada** [comment]
Comments addressed and PR updated.

**Summary of changes**:
1) Introduced xgb_train.py workflow for XGBoost training and renamed existing train.py to dl_train.py to remove ambiguity.
2) New scenario "Credit Risk" that joins 6 datasets from 4 TDPs and trains using XGBoost with DP. Here, one of the TDPs (Bank A) also acts as the TDC, to reflect a more realistic scenario of collaboration.
3) Moved parsers/contructors for model, dataset, eval and loss configs into utilities
4) Minor change in policy.rego to allow single TDC container (if base model is being instantiated inside CCR)
5) Included 2 alternate model_config samples for COVID, one that trains using XGBoost, another that packages model as a SafeTensor file and not the default ONNX.
6) Updated ReadMes: i) Added mermaid diagrams to illustrate config creation for training. ii) Added info about datasets used in scenarios. iii) Modified main readme scenarios table. iv) Renamed ACR and Contract Service identifiers.
7) Added parameter "provider" in Join config to differentiate from dataset name, since single provider may have more than one dataset

**To be done (in subsequent PRs)**:
1) "Create your own Scenario" guide and boilerplate automation
2) Classical ML training support
3) LLM Finetuning support
4) iSPIRT Contract Service 24*7 active demo, to provide users pre-signed contract sequence numbers for quick demos.
5) Optimize configurations / adopt existing standards (eg. Hydra / KubeFlow)

**Review by kapilvgit — APPROVED** [review]
